crypto/aria/aria.c

```diff
@@ -7,15 +7,834 @@
+        (TMP) = (X);                                \
+        (TMP2) = rotr32((TMP), 8);                  \
+        (Y) = (TMP2) ^ rotr32((TMP) ^ (TMP2), 16);  \
+    } while(0)
+
+void aria_encrypt(const unsigned char *in, unsigned char *out,
+                  const ARIA_KEY *key) 
+{
+    register uint32_t reg0, reg1, reg2, reg3;
+    int Nr;
+
+    const uint8_t *rk = (const uint8_t*)key->rd_key;
+
+    if (in == NULL || out == NULL || key == NULL) {
+        return;
+    }
+
+    Nr = key->rounds;
+
+    if (Nr != 12 && Nr != 14 && Nr != 16) {
+        return;
+    }
+
+    reg0 = GET_U32_BE(in, 0);
+    reg1 = GET_U32_BE(in, 1);
+    reg2 = GET_U32_BE(in, 2);
+    reg3 = GET_U32_BE(in, 3);
+
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+    ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+    rk += 16;
+
+    if (Nr > 12) {
+        ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+        ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+        rk += 16;
+
+        ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+        ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+        rk += 16;
+
+        if (Nr > 14) {
+            ARIA_SUBST_DIFF_EVEN(reg0, reg1, reg2, reg3);
+            ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+            rk += 16;
+
+            ARIA_SUBST_DIFF_ODD(reg0, reg1, reg2, reg3);
+            ARIA_ADD_ROUND_KEY(rk, reg0, reg1, reg2, reg3);
+            rk += 16;
+        }
+    }
+    
+    out[ 0] = (uint8_t)(X1[GET_U8_BE(reg0, 0)]     ) ^ rk[ 0];
+    out[ 1] = (uint8_t)(X2[GET_U8_BE(reg0, 1)] >> 8) ^ rk[ 1];
+    out[ 2] = (uint8_t)(S1[GET_U8_BE(reg0, 2)]     ) ^ rk[ 2];
+    out[ 3] = (uint8_t)(S2[GET_U8_BE(reg0, 3)]     ) ^ rk[ 3];
```

Comment:
And of course if 32-bit rk elements are stored in such way that they don't have to be picked byte by byte, this would have to look different...